python/pyspark/pandas/series.py

```diff
@@ -5228,22 +5228,128 @@ def asof(self, where: Union[Any, List]) -> Union[Scalar, "Series"]:
             where = [where]
         index_scol = self._internal.index_spark_columns[0]
         index_type = self._internal.spark_type_for(index_scol)
+
+        # e.g where = [10, 20]
+        # In the comments below , will explain how the dataframe will look after transformations.
+        # e.g pd.Series([2, 1, np.nan, 4], index=[10, 20, 30, 40], name="Koalas")
+
+        column_prefix_constant = "col_"
         cond = [
-            F.max(F.when(index_scol <= SF.lit(index).cast(index_type), self.spark.column))
-            for index in where
+            F.when(
+                index_scol <= SF.lit(index).cast(index_type),
+                F.struct(
+                    F.lit(column_prefix_constant + str(index) + "_" + str(idx)).alias("identifier"),
+                    self.spark.column.alias("col_value"),
+                ),
+            ).alias(column_prefix_constant + str(index) + "_" + str(idx))
```

Comment:
I think we can simply use `F.last` with `ignorenulls=True` instead of `F.max` ??
(https://spark.apache.org/docs/3.1.1/api/python/reference/api/pyspark.sql.functions.last.html)

e.g.
```python
        cond = [
            F.last(F.when(index_scol <= SF.lit(index).cast(index_type), self.spark.column), ignorenulls=True)
            for index in where
    ]
```

It returns the last non-null value from given column.

Reply:
Yes @itholic , this is working (since __index_level_0__) is sorted.  However , test case with psser.asof([25, 25]) , ambiguous  of duplicate cols in psdf = ps.DataFrame(sdf) . Therefore , in order to  pass above test case , 
below is the change. 
```python

cond = [
            F.last(
                F.when(index_scol <= SF.lit(index).cast(index_type), self.spark.column),
                ignorenulls=True,
            ).alias(column_prefix_constant + str(index) + "_" + str(idx))
            for idx, index in enumerate(where)
        ]

```
Then 
```python
with ps.option_context("compute.default_index_type", "distributed", "compute.max_rows", 1):
            psdf = ps.DataFrame(sdf)  # type: DataFrame
            df = pd.DataFrame(psdf.transpose().values, columns=[self.name], index=where)
            return df[df.columns[0]]

```

Reply:
I think we might want to return the pandas-on-Spark Series rather than pandas Series, and should leverage the pandas `Series` directly when only the `where` has duplicate item. (and seems like `column_prefix_constant` is not necessary?)

So, how about this ??

```python
        cond = [
            F.last(
                F.when(index_scol <= SF.lit(index).cast(index_type), self.spark.column),
                ignorenulls=True,
            )
            for index in where
        ]
```

Then

```python
        # The data is expected to be small so it's fine to transpose/use default index.
        with ps.option_context("compute.default_index_type", "distributed", "compute.max_rows", 1):
            if len(where) == len(set(where)):
                psdf: DataFrame = DataFrame(sdf)
                psdf.columns = pd.Index(where)
                return first_series(psdf.transpose()).rename(self.name)
            else:
                # If `where` has duplicate items, leverage the pandas directly
                # since pandas API on Spark doesn't support the duplicate column name.
                pdf: pd.DataFrame = sdf.limit(1).toPandas()
                pdf.columns = pd.Index(where)
                return first_series(DataFrame(pdf.transpose())).rename(self.name)
```

??

Reply:
@itholic  Yes i think this is good suggestion.  Please let me know , if I update this PR , or you are planning to create new PR with the suggested code changes.

Reply:
It's okay to just update here! :-)